internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,117 @@
 package buffer
 
 import (
-	"bytes"
+	"fmt"
 )
 
+// InMemoryWriteBuffer implements WriteBuffer and is stored in memory.
 type InMemoryWriteBuffer struct {
-	// Holds the data to be written to GCS. At any time, Buffer holds the data for
-	// 2 GCS write calls.
-	buffer *bytes.Buffer
+	// Holds the incoming data from kernel write calls.
+	currentBuffer []byte
+	// [ currentBufferStartOffset, currentBufferEndOffset) is the range of file offset for which the
+	// data is currently stored in currentBuffer.
+	currentBufferStartOffset int64
+	currentBufferEndOffset   int64
+
+	// Holds the data that is sent to go storage client for upload. Data in this buffer
+	// will persist until the entire data is successfully uploaded to GCS.
+	flushedBuffer []byte
+	// [ flushedBufferStartOffset, flushedBufferEndOffset) is the range of file offset for which the
+	// data is currently stored in flushedBuffer.
+	flushedBufferStartOffset int64
+	flushedBufferEndOffset   int64
+
+	// Holds the last offset to which data is written.
+	fileSize int64
 }
 
 // CreateInMemoryWriteBuffer creates a buffer with InMemoryWriteBuffer.buffer
 // set to nil. Memory is allocated to the buffer when the first write call comes.
 // This avoids unnecessarily bloating GCSFuse memory consumption.
 //
-// To allocate memory to the buffer, use WriteBuffer.InitializeBuffer.
+// To allocate memory to the buffer, use WriteBuffer.Initialize.
 func CreateInMemoryWriteBuffer() *InMemoryWriteBuffer {
 	b := &InMemoryWriteBuffer{}
 	// TODO: set mtime attribute.
 	return b
 }
 
-func (b *InMemoryWriteBuffer) InitializeBuffer(sizeInMB int) {
-	ChunkSize = sizeInMB * MiB
-	if b.buffer == nil {
-		b.buffer = bytes.NewBuffer(make([]byte, 0, 2*sizeInMB*MiB))
+func (b *InMemoryWriteBuffer) Initialize(sizeInMB int) {
+	if ChunkSize == 0 {
+		ChunkSize = int64(sizeInMB * MiB)
+	}
+	if b.currentBuffer == nil {
+		b.currentBuffer = make([]byte, 0, ChunkSize)
+		b.currentBufferEndOffset = ChunkSize
+	}
+	if b.flushedBuffer == nil {
+		b.flushedBuffer = make([]byte, 0, ChunkSize)
+	}
+}
+
+func (b *InMemoryWriteBuffer) WriteAt(content []byte, receivedContentStartOffset int64) error {
+	dataSize := int64(len(content))
+	receivedContentEndOffset := receivedContentStartOffset + dataSize
+
+	// If receivedContentStartOffset is not in the range of offsets stored in current
+	// buffer, trigger temp file flow for writes.
+	if receivedContentStartOffset < b.currentBufferStartOffset ||
+		b.currentBufferEndOffset < receivedContentStartOffset {
+		// TODO: finalise write and trigger temp file flow.
+		return fmt.Errorf(NonSequentialWriteError)
+	}
+
+	// If data can be completely written to the current buffer block, write it.
+	if receivedContentEndOffset <= b.currentBufferEndOffset {
+		return b.copyDataToBuffer(receivedContentStartOffset, receivedContentEndOffset, content)
+	} else {
+		// Below logic is written with the assumption that the data received in a
```

Comment:
Can we put some kind of check to make sure of this, rather than assuming it?

Reply:
This will always hold true because we never receive more than 1 MiB data in 1 kernel write call. I don't think there's a need for this check. Let me know what you think.

Reply:
This is a generic utility, which in this case will get its data from kernel, so that assumption will never be tested in this use case. But as this is a public interface, it can really be used by any other caller. So, let's either (a) state (int header), check (in implementation that that case isn't supported) and test (as a negative unit test) that assumption, or (b) not make that assumption, add check and test.